center/src/main/java/com/microsoft/hydralab/center/controller/StorageController.java

```diff
@@ -22,14 +42,111 @@
+                             @RequestParam("fileUri") String fileUri) {
+        String storageToken = request.getHeader("Authorization");
+        if (storageToken != null) {
+            storageToken = storageToken.replaceAll("Bearer ", "");
+        } else {
+            return Result.error(HttpStatus.UNAUTHORIZED.value(), "Invalid visit with no auth code");
+        }
+        if (!storageTokenManageService.validateAccessToken(storageToken)) {
+            return Result.error(HttpStatus.UNAUTHORIZED.value(), "Unauthorized, error access token for storage actions.");
+        }
+        if (!LogUtils.isLegalStr(fileUri, Const.RegexString.STORAGE_FILE_REL_PATH, false)) {
+            throw new HydraLabRuntimeException(HttpStatus.BAD_REQUEST.value(), "Invalid file path!");
+        }
+        File file = new File(Const.LocalStorageURL.CENTER_LOCAL_STORAGE_ROOT + fileUri);
+        File parentDirFile = new File(file.getParent());
+        if (!parentDirFile.exists() && !parentDirFile.mkdirs()) {
+            return Result.error(HttpStatus.INTERNAL_SERVER_ERROR.value(), "mkdirs failed!");
+        }
+
+        InputStream inputStream;
+        try (FileOutputStream fileOutputStream = new FileOutputStream(file)) {
+            inputStream = uploadedFile.getInputStream();
+            IOUtils.copy(inputStream, fileOutputStream);
+        } catch (IOException e) {
+            return Result.error(HttpStatus.INTERNAL_SERVER_ERROR.value(), "upload file failed!");
+        }
+        return Result.ok(fileUri);
+    }
+
+    // used by center/agent
+    @PostMapping(Const.LocalStorageURL.CENTER_LOCAL_STORAGE_DOWNLOAD)
+    public void postDownloadFile(HttpServletRequest request,
+                                 HttpServletResponse response,
+                                 @RequestParam("fileUri") String fileUri) {
+        String storageToken = request.getHeader("Authorization");
+        if (storageToken != null) {
+            storageToken = storageToken.replaceAll("Bearer ", "");
+        } else {
+            throw new HydraLabRuntimeException(HttpStatus.UNAUTHORIZED.value(), "Invalid visit with no auth code");
+        }
+        if (!storageTokenManageService.validateAccessToken(storageToken)) {
+            throw new HydraLabRuntimeException(HttpStatus.UNAUTHORIZED.value(), "Unauthorized, error access token for storage actions.");
+        }
+        if (!LogUtils.isLegalStr(fileUri, Const.RegexString.STORAGE_FILE_REL_PATH, false)) {
+            throw new HydraLabRuntimeException(HttpStatus.BAD_REQUEST.value(), "Invalid file path!");
+        }
+
+        copyDownloadStreamToResponse(fileUri, response);
+    }
+
+    // for front end to download file
+    @GetMapping("/api/storage/local/download/**")
+    public void getDownloadFile(HttpServletRequest request,
+                                HttpServletResponse response,
+                                @QueryParam("token") String token) {
```

Comment:
Should we put the token in header to follow a common practice?

Reply:
This API is used for front end, and similar API of AZURE BLOB is the conjunct of URL and token string as query string.
If change to header, frontend may involve extra modification